tech/firewall-rule-format.md

```diff
@@ -0,0 +1,574 @@
+      },
+      {
+        "kind": "compute#firewall",
+        "id": "6979850877904078814",
+        "creationTimestamp": "2018-12-04T04:55:13.187-08:00",
+        "name": "default-allow-ssh",
+        "description": "Allow SSH from anywhere",
+        "network": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/networks/default",
+        "priority": 65534,
+        "sourceRanges": [
+          "0.0.0.0/0"
+        ],
+        "allowed": [
+          {
+            "IPProtocol": "tcp",
+            "ports": [
+              "22"
+            ]
+          }
+        ],
+        "direction": "INGRESS",
+        "logConfig": {
+          "enable": false
+        },
+        "disabled": false,
+        "selfLink": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/firewalls/default-allow-ssh"
+      }
+    ]
+
+Since GCP returns one JSON object per firewall rule while Azure groups
+firewall rules under a network security group, the Azure cloud plugin
+should generate multiple firewall rule records, one for each firewall
+rule in a network security group in order to maintain parity with the
+records of GCP firewall rules.
+
+Further the `com` (common notation) bucket of the generated records
+should be populated with generic attributes that are common in both
+Azure and GCP firewall rules.
+
+From the next section onwards, we will compare a single firewall rule,
+i.e., a single item in the `security_rules` bucket in Azure with a
+single `compute#firewall` document in GCP for various types of firewall
+rules.
+
+
+Multiple Ranges
+---------------
+
+There are primarily three important pieces of information we need to
+analyze in a firewall rule to understand if the firewall rule is
+potentially weak:
+
+  - Direction: We will focus on inbound rules first. Azure calls it
+    `Inbound`. GCP calls it `INGRESS`. In the common notation, we will
+    write it as `in`.
+
+  - Access: Rules can either deny access or allow access. We will focus
+    only on those that allow access.
+
+  - Source address range: We need to understand how they look in the
+    JSON data when multiple ranges are defined. We want to find those
+    rules that have `*` or `0.0.0.0/0` or any other pattern that
+    indicates *all* source address ranges.
+
+  - Destination port range: We need to understand how they look in the
+    JSON data when multiple ranges are defined. If a range includes
+    critical ports like 22 (SSH), 3389 (RDP), etc. then we want to check
+    if they are allowed to the all source address ranges.
+
+
+### Azure
+
+Here is an example Azure security rule that allows multiple source
+address ranges to connect to multiple destination port ranges:
+
+    {
+      "id": "/subscriptions/a8daa361-b951-46b5-8e9d-b375f2ffe9fd/resourceGroups/rg1/providers/Microsoft.Network/networkSecurityGroups/vm1-nsg/securityRules/Port_Multi_Range",
+      "protocol": "*",
+      "source_port_range": "*",
```

Comment:
I feel that to compare the Azure NSG rule with a GCP firewall, the example in this line should also be restricted to a particular protocol like "tcp" or "udp" like it is in the GCP firewall object. A point to note from the documentation of [Azure](https://docs.microsoft.com/en-us/rest/api/virtualnetwork/networksecuritygroups/listall#securityruleprotocol) and [GCP](https://cloud.google.com/compute/docs/reference/rest/v1/firewalls/list#response-body) is that Azure supports fewer protocol values (Tcp, Udp, *) than GCP does (tcp, udp, icmp, esp, ah, ipip, sctp).

I also don't understand why GCP does not have any restrictions for source ports or destination IPs. If the design was such that a GCP firewall rule is applied only to one GCP compute, it would have made sense to omit a destination IP range. But GCP firewalls seem to be getting applied to the network and not the compute.

Reply:
Good points. For the first point, I will create a separate section to explore the possible values of the protocols. In that section, I would provide two similar examples for Azure and GCP with specific protocols like `tcp` and `udp` appearing in the Azure JSON object. In this section, I did not think about taking a closer look at the protocol, so I went with the default value for the protocol which in the Azure protocol is "Any". It appears as `*` in the JSON object.

<img width="225" alt="Screenshot of Azure firewall rule protocol option" src="https://user-images.githubusercontent.com/229488/54495708-5706a800-490c-11e9-9784-c44ae205e95d.png">

In Google Cloud Console, though, there was no option for selecting protocol as "Any". I had to enter a port range under "tcp" and/or "udp".

<img width="500" alt="Screenshot of GCP firewall rule protocol option" src="https://user-images.githubusercontent.com/229488/54495716-6128a680-490c-11e9-9fdd-776bd1f3fd16.png">

There is an "Allow All" option though that seems to allow connections to all ports under all protocols which I would be exploring later.

Reply:
About the second point, the source ports are usually client ports which almost always happen to be ephemeral ports. I believe they thought that since ephemereal ports always keep changing, there would be little value in adding support for them in firewall rules.

The same holds true for destination IP addresses too. The destination IP addresses would keep changing because there are no static IP addresses in the cloud. Perhaps they do not want the users to rely on the IP addresses of the VMs while configuring firewall rules.

Reply:
This makes sense. Thank you.

Reply:
Addressed the first point in pull request #4 by explaining that we are not analyzing the difference in protocol format in this section.